### PR TITLE
Add notebook for Fitting DPMM with Symmetric Dirichlet Prior Using pSGLD

### PR DESCRIPTION
**Pull request by ykkawana**
I added a jupyter notebook of fitting Dirichlet Process Mixture Model with Symmetric Dirichlet Prior Using Preconditioned Stochastic Gradient Langevin Dynamics for clustering a toy dataset.

### CODE REVIEW

**Comment by ykkawana** [comment]
I signed it!

**Comment by ykkawana** [comment]
I signed it!

**Comment by ykkawana** [comment]
I signed it!

**Comment by ykkawana** [comment]
I signed it!

**Comment by ykkawana** [comment]
I signed it

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb [hunk 1]
```diff
@@ -0,0 +1,975 @@
...
+      ]
+    },
+    {
+      "metadata": {
+        "id": "vsGdTAPr2FtM",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "In this notebook, we will demonstrate how to cluster a large number of samples and infer the number of clusters simultaneously by fitting Dirichlet Process Mixture of Gaussian distribution with Symmetric Dirichlet Prior using Preconditioned Stochastic Gradient Langevin Dynamics (pSGLD). "
```
Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: I might break this up in a few lines, by saying something like:

"by fitting a Dirichlet Process Gaussian Mixture Model. We use Preconditioned SGLD for inference."

The statement about what prior is being used can be stated below in the code / notes (and in fact I might remove the prior from the title itself).

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb [hunk 2]
```diff
@@ -0,0 +1,975 @@
...
+        "from __future__ import absolute_import\n",
+        "from __future__ import division\n",
+        "from __future__ import print_function\n",
+        "\n",
+        "import tensorflow as tf\n",
+        "import tensorflow_probability as tfp\n",
+        "import numpy as np\n",
+        "import matplotlib.pyplot as plt; plt.style.use('ggplot')\n",
+        "import matplotlib.cm as cm\n",
+        "from tqdm import tqdm\n",
```
Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: I don't know if we'll guarantee tqdm being installed, so I might leave this out.

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb [hunk 3]
```diff
@@ -0,0 +1,975 @@
...
+      "source": [
+        "# For reproducibility\n",
+        "random = np.random.RandomState(seed=45)\n",
+        "tf.set_random_seed(76)\n",
+        "\n",
+        "# Precision\n",
+        "dtype = np.float64\n",
+        "\n",
+        "# Number of training samples\n",
+        "samples_num = 50000\n",
```
Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: num_samples

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb [hunk 4]
```diff
@@ -0,0 +1,975 @@
...
+    {
+      "metadata": {
+        "id": "ud3XnrsmCHrb",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "# For reproducibility\n",
+        "random = np.random.RandomState(seed=45)\n",
```
Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: I'd use rng (just to distinguish from a possible random module).

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb [hunk 5]
```diff
@@ -0,0 +1,975 @@
...
...8FE+T37RK77e5Kcd\n0ZrNn/mEbJSSeNbUPWz8rOQ+3zplhUedMrkvP5GJ1I2jYdz9Ivn7ywdh7H3w1nyMcUmS5cQsmdcV\nuX3h+zJC4wG17AQpj6ssxb67auwVJTTQlq8Xg/MIy5fV3tJai/lsDfbtV3FldPW1+szbjV08XzzE\ngn2SaOdpSVrktDtdPF+S61q2l/V3kKQvDx0cCVUNx1ccz2ccjJwd8niDoymwcK6E6uOTcCUk4z5b\nBEVnsRvXS2tXA/alWRJpWbwAb8je/3R+JZJgpVRu3PiqXSMv/t0N0lpWUZSah2rLXwRK082uiD6y\nN1M6qmQrTJuzCzvzMRg3HtO2U4CuuI2Nk/awQ36OyRwEV10LGGntOnw0xCfCnh0+b334bZLk5Wn9\nWpAPJ36QGndPu1el8iSnw7EjEJ8Ey96SOvicXdCiHXb24xDdUFQBt2+CyEhsl6ukZW+/Id6eAUDg\ndyF3l+RirPkA03cQrsyBgd+D86RevXocP3488Lt7Dq1lazqXgj55uI4vnMcGVaMtr557TacUT8la\ni92fC8Zp2YmUxnk8LBOXhJ2YJbXxKWmwbhV21bvirTeOg8ZNpNbd7YYbb5c1+NXvyev+HqiK15SP\nKxLczgQoNlE6xXkxEkHJ3SF6/avfl5r2Uffgyhwomv/xSbgTm0HDxiJHu2geAHbxArmPWOzenZI4\n+dIs33fBqZaQCLmF5Auf1V5WJElRlJqDGveaTmkh/dxdshZ7w23e1p2iVDcVDuRj5zwj/bcXzcPe\nOBoG/Qy6fCK65S8+Lcp0bseIL57ne158LfXK+nD8+ws/zlDG7RfZ8Bj2G0bLOnu602Bn93ZY9b58\nvh8uAcCO/i9vuaPJ24PpeTXus2dF/nfhK3DTGKy7CLImAkaOvet+p52vowHfNN2nUngeofKKqNgp\nihI6aEJdDafUOuPkNMmKf/tVyN0tBmJiFmBkjX34KFmr9SifzX9eyt76DYW77se3hutn5AE6dvc9\nb5JUvmGPuATnh1dE+ZrqJDULvs+SV6W1679nySRstWPYO/eWCoUPl2DXrZREx78+hH3iYTGur74A\nX38l0ZO35sHB/YCR+2nkqc0qluBmkUnd+STQqRKdooQVl+D/zKGHx6uySamBmuPdMyWVyrph2mRc\nk7JEmnTsfRJ6
```
Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: I'd switch calls to get_variable to creating a  tf.compat.v2.Variable(...). This is because get_variable will not be used in TF 2.0 (so this is in order to make this notebook forwards compatible). Here and below.

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: Might name this to max_cluster_num? And then change the comment to be something like an upperbound on `K`?

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: Maybe I am missing this (since it is not rendering), but it would be nice to say what the result of pSGLD is (i.e. distinguish between a traditional optimizer, vs.  one which is a MC scheme to generate posterior samples).

**Comment by junpenglao** [comment]
I think the file is corrupted (which is why it is not rendered on Github).

**Comment by ykkawana** [comment]
@srvasude @junpenglao Thank you for your comments, I found the unintended line break in the file causing the json parsing error. I've fixed this error first and now you can see the rendering. Sorry for the inconvenience.
I will work on the other comments too.

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb (hunk 1 above)
Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: I broke the part into two lines as suggested and I dropped the "symmetric dirichlet prior" from title and file names.

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb (hunk 2 above)
Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: Removed tqdm.

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb (hunk 3 above)
Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: I replaced all samples_num with num_samples.

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb (hunk 4 above)
Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: I replaced random with rng as:
`rng = np.random.RandomState(seed=45)`

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb (hunk 5 above)
Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: First, tf.compat.v2.Variable wasn't found so I simply used tf.Variable as it seems to be supported in TF 2.0. Would this be ok?

Relating to the compatibility issue with TF 2.0, I spotted lines relating with tf.data.* gives waning:
`WARNING:tensorflow:From <ipython-input-9-46943ed918c7>:2: DatasetV1.make_one_shot_iterator (from tensorflow.python.data.ops.dataset_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use for ... in dataset: to iterate over a dataset. If using tf.estimator, return the Dataset object directly from your input function. As a last resort, you can use tf.compat.v1.data.make_one_shot_iterator(dataset).` 
So I used tf.compat.v1.data as suggesed in the warning.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: Renamed initial_cluster_num to max_cluster_num, added comment on top of it to clarify  max_cluster_num is an upperbound on K.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_with_Symmetric_Dirichlet_Prior_Using_pSGLD.ipynb]: Added the additional result on fitting the model using RMSProp as traditional optimizer. I mentioned the processing time, visualized the clustering result and mixture weights of the components for easy comparison.
I also added the experiment on uncertainty to distinguish pSGLD from MAP inference of traditional optimizer.

File: tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb [hunk 6]
```diff
@@ -0,0 +1,1407 @@
...
...5GJ1I2jYdz9Ivn7ywdh7H3w1nyMcUmS5cQsmdcV\nuX3h+zJC4wG17AQpj6ssxb67auwVJTTQlq8Xg/MIy5fV3tJai/lsDfbtV3FldPW1+szbjV08XzzE\ngn2SaOdpSVrktDtdPF+S61q2l/V3kKQvDx0cCVUNx1ccz2ccjJwd8niDoymwcK6E6uOTcCUk4z5b\nBEVnsRvXS2tXA/alWRJpWbwAb8je/3R+JZJgpVRu3PiqXSMv/t0N0lpWUZSah2rLXwRK082uiD6y\nN1M6qmQrTJuzCzvzMRg3HtO2U4CuuI2Nk/awQ36OyRwEV10LGGntOnw0xCfCnh0+b334bZLk5Wn9\nWpAPJ36QGndPu1el8iSnw7EjEJ8Ey96SOvicXdCiHXb24xDdUFQBt2+CyEhsl6ukZW+/Id6eAUDg\ndyF3l+RirPkA03cQrsyBgd+D86RevXocP3488Lt7Dq1lazqXgj55uI4vnMcGVaMtr557TacUT8la\ni92fC8Zp2YmUxnk8LBOXhJ2YJbXxKWmwbhV21bvirTeOg8ZNpNbd7YYbb5c1+NXvyev+HqiK15SP\nKxLczgQoNlE6xXkxEkHJ3SF6/avfl5r2Uffgyhwomv/xSbgTm0HDxiJHu2geAHbxArmPWOzenZI4\n+dIs33fBqZaQCLmF5Auf1V5WJElRlJqDGveaTmkh/dxdshZ7w23e1p2iVDcVDuRj5zwj/bcXzcPe\nOBoG/Qy6fCK65S8+Lcp0bseIL57ne158LfXK+nD8+ws/zlDG7RfZ8Bj2G0bLOnu602Bn93ZY9b58\nvh8uAcCO/i9vuaPJ24PpeTXus2dF/nfhK3DTGKy7CLImAkaOvet+p52vowHfNN2nUngeofKKqNgp\nihI6aEJdDafUOuPkNMmKf/tVyN0tBmJiFmBkjX34KFmr9SifzX9eyt76DYW77se3hutn5AE6dvc9\nb5JUvmGPuATnh1dE+ZrqJDULvs+SV6W1679nySRstWPYO/eWCoUPl2DXrZREx78+hH3iYTGur74A\nX38l0ZO35sHB/YCR+2nkqc0qluBmkUnd+STQqRKdooQVl+D/zKGHx6uySamBmuPdMyWVyrph2mRc\nk7JEmnTsfRJ6b9REHle+A6vekTdb9S5Ex0ir15dmwfWjxJAsekVC8hvX+05ckCePsQl
```
Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: I'd change this to: "For an ideal Dirichlet Mixture Model, $K$ is set to $\\infty$.  However, it is known..."

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Remove the np.zeros() line? (just use initial_value=np.random.uniform(....))

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: instead of "the gradient", I might say "the update is sampled as"

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Add space `rv_loc.log_prob(loc) / num_samples`. Here and below.

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Change to tf.compat.v1.placeholder.

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Extra backslash.

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: unnormalized_posterior. Also this line seems to have extra backslashes rendering.

Comment (srvasude) [inline tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: We use entropy as a measure of uncertainty.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Modified the line as suggested.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Removed np.zeros() as initialization and instead use np.random.uniform directly.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Modified the line as suggested.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: I added spaces around all operators including that line.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Modified as suggested.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Modified `unnormalized_posterior_of_z` to `unnormalized_posterior` and `posterior_of_z` to `posterior`.
I also tried to remove backlashes for breaking the line but this caused error. I keep it as it is now.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Took out a backlash.

Comment (ykkawana) [reply tensorflow_probability/examples/jupyter_notebooks/Fitting_DPMM_Using_pSGLD.ipynb]: Modified the line as suggested.

**Comment by ykkawana** [comment]
@srvasude Thank you for your review. I added the simple table of contents but is this what you had expected? I tried to add a table of contents with links to each section. But the link doesn't work when it's displayed in GitHub so I made the table of contents without links.

**Comment by srvasude** [comment]
Hi! This should be checked in very soon. As follow ups, some things that might be nice to have would be for the notebook to use TF2, and use the new-fangled JointDistribution: https://github.com/tensorflow/probability/blob/0cfc908a6a572914aba2de73153eacbdde93651b/tensorflow_probability/python/distributions/joint_distribution.py

**Comment by ykkawana** [comment]
@srvasude Thank you for all the reviews you made and follow up. I'm happy to see the commit has merged! Will work on update for TF2 too.
